### PR TITLE
Vyper Compiler Notes

### PR DESCRIPTION
**Pull request by jacqueswww**
### Cute Animal Picture
`^.^`

Put a link to a cute animal picture inside the parenthesis-->

### CODE REVIEW

File: docs/compiler-notes.rst [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+Introduction
+~~~~~~~~~~~~
+
+As Vyper targets Smart Contract development - there are some notable differences. This section
+aims to highlight the important parts.
+
+Types
+~~~~~
+
+One of the key difference between Python and Vyper is that Python usese dynamic typing; and Vyper
```
Comment (arjunaskykok) [inline docs/compiler-notes.rst]: Typo: "usese".

Comment (fubuloubu) [inline docs/compiler-notes.rst]:
```suggestion
One of the key difference between Python and Vyper is that Python uses dynamic typing; and Vyper
```

File: docs/compiler-notes.rst [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+uses static Typing. Type in Vyper are declared using Python Annotations e.g. `a: auint256`.
```
Comment (fubuloubu) [inline docs/compiler-notes.rst]:
```suggestion
uses static Typing. Type in Vyper are declared using Python Annotations e.g. `a: uint256`.
```

File: docs/compiler-notes.rst [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+- shifting operators not supported. Use `shift` insetad.
+- floor division operator `//` is not supported. Use `floor` and `ceil` instead, to make rounding
+  operations clear.
+
+Statements
+~~~~~~~~~~
+
+*Simple Statements*
+
+**Airthmetic Conversions**
```
Comment (fubuloubu) [inline docs/compiler-notes.rst]:
```suggestion
**Arithmetic Conversions**
```

File: docs/compiler-notes.rst [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+
+Vyper enforces airthmetic conversions, to be explicit vs. python which uses implicit type conversion.
```
Comment (fubuloubu) [inline docs/compiler-notes.rst]:
```suggestion
Vyper enforces arithmetic conversions, to be explicit vs. python which uses implicit type conversion.
```

File: docs/compiler-notes.rst [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+
+Vyper has support for integer, decimal, string and byte python literals.
+- If to Literals are found in an expression they are constant folded.
+  - `10*10` will be converted to `100`
+  - However `10*10*a` will not be converted.
+- See above artihmetic onversion rules.
+- There are limitations to byte & string literals based on their usage.
+  - e.g. A string literal of size 100 can't be assigned or return to an annoted size of 32.
+- To distinguish between `int128` and `uint256` literals Vyper uses the following ruleset.
+  - If literal int value is above MAXNUM (2**127 - 1), literal is a uint256.
```
Comment (fubuloubu) [inline docs/compiler-notes.rst]: Would this be a correct way of stating this?

```suggestion
  - If literal int value is above `MAXNUM` (`2**127 - 1`), the literal must be assigned to an uint256.
```

File: docs/compiler-notes.rst [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+  - If literal int value is under MINNUM (-2**127), literal is invalid.
+  - If assignment of literal to a known type (could be uint256 or int128), use appropriate type.
+  - Consequently slightly confusing error message of `Cannot assign uint256 to int128` does occur,
+    but seemed to be a reasonable trade off (don't have to convert(xxxx, uint256) on assingment).
```
Comment (fubuloubu) [inline docs/compiler-notes.rst]:
```suggestion
    but seemed to be a reasonable trade off (don't have to convert(xxxx, uint256) on assignment).
```

File: docs/compiler-notes.rst [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+
+**Comparisons / Boolean Operations*
+
+- Comparisons
+  - As long as all members of are of the same type -comparisons are supported.
```
Comment (fubuloubu) [inline docs/compiler-notes.rst]:
```suggestion
  - As long as all members of are of the same type, comparisons are supported.
```

File: docs/compiler-notes.rst [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+- Boolean `and`, `or` operations are supported.
+    - Currently not short-circuited (VIP pending)
```
Comment (fubuloubu) [inline docs/compiler-notes.rst]: Link to VIP?

File: docs/compiler-notes.rst [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+
+Vyper does not support `del ..` to clear values from storage use `clear()` instead.
+
+** Import Statements **
+
+Import statements are supported for use with import interfaces, and do not import functions.
+
+** Assert Statements **
+
+`assert` is supported and uses the `REVERT` opcode. To `INVALID` opcode, use assert <expre>, UNREACHABLE
```
Comment (fubuloubu) [inline docs/compiler-notes.rst]:
```suggestion
`assert` is supported and uses the `REVERT` opcode. To use `INVALID` opcode, use assert <expre>, UNREACHABLE
```

File: docs/compiler-notes.rst [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+
+`raise` is supported, uses the `REVERT` opcode to throw an exception (with reason string smaller 
+than 32 bytes)
+
+** Return Statements **
+
+- `return` statements are supported.
+- There is a checker in place to ensure return statements are not missing in branches.
+
+*Compound Statements*
```
Comment (fubuloubu) [inline docs/compiler-notes.rst]: Missing stuff here

File: docs/compiler-notes.rst [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+
+**Functions**
+
+- Vyper support default arguments for both public and private functions.
```
Comment (fubuloubu) [inline docs/compiler-notes.rst]: What happens with default arguments in private functions? Do we have test cases for this?

**Comment by fubuloubu** [comment]
@jacqueswww can you rebase and apply some of the changes from above?

**Comment by fubuloubu** [comment]
Closing this for now, but leave this as a reference. @iamdefinitelyahuman has been doing work to merge these sorts of notes into README.md files under various subfolders of the compiler. That work should continue, and could leverage some of these notes to finish it.

Part of work on #1618

**Pull request closed without merging**